src/buffer/out/UTextAdapter.cpp

```diff
@@ -0,0 +1,328 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "precomp.h"
+#include "UTextAdapter.h"
+
+#include "textBuffer.hpp"
+
+struct RowRange
+{
+    til::CoordType begin;
+    til::CoordType end;
+};
+
+constexpr size_t& accessLength(UText* ut) noexcept
+{
+    return *std::bit_cast<size_t*>(&ut->p);
+}
+
+constexpr RowRange& accessRowRange(UText* ut) noexcept
+{
+    return *std::bit_cast<RowRange*>(&ut->a);
+}
+
+constexpr til::CoordType& accessCurrentRow(UText* ut) noexcept
+{
+    return ut->b;
+}
+
+// An excerpt from the ICU documentation:
+//
+// Clone a UText. Much like opening a UText where the source text is itself another UText.
+//
+// A shallow clone replicates only the UText data structures; it does not make
+// a copy of the underlying text. Shallow clones can be used as an efficient way to
+// have multiple iterators active in a single text string that is not being modified.
+//
+// A shallow clone operation must not fail except for truly exceptional conditions such
+// as memory allocation failures.
+//
+// @param dest   A UText struct to be filled in with the result of the clone operation,
+//               or NULL if the clone function should heap-allocate a new UText struct.
+// @param src    The UText to be cloned.
+// @param deep   true to request a deep clone, false for a shallow clone.
+// @param status Errors are returned here.  For deep clones, U_UNSUPPORTED_ERROR should
+//               be returned if the text provider is unable to clone the original text.
+// @return       The newly created clone, or NULL if the clone operation failed.
+static UText* U_CALLCONV utextClone(UText* dest, const UText* src, UBool deep, UErrorCode* status) noexcept
+{
+    __assume(status != nullptr);
+
+    if (deep)
+    {
+        *status = U_UNSUPPORTED_ERROR;
+        return dest;
+    }
+
+    dest = utext_setup(dest, 0, status);
+    if (*status <= U_ZERO_ERROR)
+    {
+        memcpy(dest, src, sizeof(UText));
+    }
+
+    return dest;
+}
+
+// An excerpt from the ICU documentation:
+//
+// Gets the length of the text.
+//
+// @param ut the UText to get the length of.
+// @return the length, in the native units of the original text string.
+static int64_t U_CALLCONV utextNativeLength(UText* ut) noexcept

```

Comment:
should these functions return a failure if the underlying buffer was mutated since the utext adapter was created? Or is that not related to the purpose of these guys?

Reply:
I feel like that this would provide a false sense of security that people might end up relying on. Unlocking the console at all during a search is likely a bug or at least a bad idea because it's pretty error prone (not all state relevant to searching is contained in TextBuffer after all - selections for instance). I think we could make it an assertion? Not sure if it should be `assert()` or `WI_ASSERT()`. Probably the former, because searching without holding the lock is still somewhat predictable.